File: comparison/dat_parser.py
# -*- coding: utf-8 -*-
"""פענוח קובץ DAT (ממשק לאוצר).

הקובץ מופרד בתו '~' ומקודד בקידוד עברי ישן (IBM/DOS-862).
בשלב זה מפוענחות רק שורות שמתחילות ב-9050 (תקופות עבודה).

מבנה שורת 9050 (לפי סדר השדות בקובץ):
    0: קוד רשומה (9050)
    1: לא ידוע
    2: לא ידוע
    3: מספר זהות
    4: לא ידוע
    5: קוד סוג תקופה
    6: תאריך התחלה (DDMMYYYY)
    7: תאריך סיום (DDMMYYYY)
    8: אורך שירות בחודשים כפול 100 (למשל 01200 = 12 חודשים)
    9: קוד סוג זכויות
   10: היקף משרה כפול 1000 (למשל 0667 = 0.667)
"""
from dataclasses import dataclass, field

ENCODINGS = ("cp862", "iso-8859-8", "utf-8")
MIN_9050_FIELDS = 11


@dataclass
class DatPeriod:
    """תקופת עבודה אחת משורת 9050."""
    id_number: str          # מספר זהות מנורמל (ללא אפסים מובילים)
    sug_tkufa: int          # קוד סוג תקופה
    start: str              # DDMMYYYY
    end: str                # DDMMYYYY
    months: float           # אורך שירות בחודשים
    sug_zchuyot: int        # קוד סוג זכויות
    heikef: float           # היקף משרה (0-1)
    line_number: int        # מספר שורה בקובץ, להודעות שגיאה


@dataclass
class DatParseResult:
    periods_by_id: dict = field(default_factory=dict)   # id -> [DatPeriod]
    warnings: list = field(default_factory=list)
    errors: list = field(default_factory=list)


def normalize_id(raw):
    """נרמול מספר זהות: הסרת רווחים ואפסים מובילים (ב-PDF מודפס ללא אפס מוביל)."""
    return raw.strip().lstrip("0")


def decode_dat(data: bytes) -> str:
    """פענוח בייטים של קובץ DAT. מנסה קידוד DOS-862 (הרגיל) ואז חלופות."""
    for enc in ENCODINGS:
        try:
            return data.decode(enc)
        except UnicodeDecodeError:
            continue
    # קידוד לא מזוהה - פענוח סלחני כדי שהשורות המספריות עדיין יעבדו
    return data.decode("cp862", errors="replace")


def parse_dat_bytes(data: bytes) -> DatParseResult:
    result = DatParseResult()
    text = decode_dat(data)
    for line_no, line in enumerate(text.splitlines(), start=1):
        if not line.startswith("9050~"):
            continue
        fields = line.rstrip().split("~")
        if len(fields) < MIN_9050_FIELDS:
            result.errors.append(
                f"שורה {line_no}: שורת 9050 עם {len(fields)} שדות בלבד "
                f"(נדרשים {MIN_9050_FIELDS}) - השורה דולגה"
            )
            continue
        try:
            period = DatPeriod(
                id_number=normalize_id(fields[3]),
                sug_tkufa=int(fields[5]),
                start=fields[6].strip(),
                end=fields[7].strip(),
                months=int(fields[8]) / 100.0,
                sug_zchuyot=int(fields[9]),
                heikef=int(fields[10]) / 1000.0,
                line_number=line_no,
            )
        except ValueError as exc:
            result.errors.append(f"שורה {line_no}: ערך לא מספרי בשורת 9050 ({exc}) - השורה דולגה")
            continue
        result.periods_by_id.setdefault(period.id_number, []).append(period)

    # בדיקת שפיות: לא אמורות להיות שתי תקופות עם אותם תאריכי התחלה וסיום לאותה ת"ז
    for id_number, periods in result.periods_by_id.items():
        seen = {}
        for p in periods:
            key = (p.start, p.end)
            if key in seen:
                result.warnings.append(
                    f'ת"ז {id_number}: נמצאו שתי תקופות עם אותם תאריכים '
                    f"({fmt_date(p.start)} - {fmt_date(p.end)}, שורות {seen[key]} ו-{p.line_number}). "
                    f"ההשוואה עבור התקופה הזו עלולה להיות שגויה."
                )
            else:
                seen[key] = p.line_number
    return result


def parse_dat_file(path: str) -> DatParseResult:
    with open(path, "rb") as f:
        return parse_dat_bytes(f.read())


def fmt_date(ddmmyyyy: str) -> str:
    """DDMMYYYY -> DD/MM/YYYY לתצוגה."""
    if len(ddmmyyyy) == 8:
        return f"{ddmmyyyy[:2]}/{ddmmyyyy[2:4]}/{ddmmyyyy[4:]}"
    return ddmmyyyy

File: src/datParser.js
/**
 * פענוח קובץ DAT (ממשק לאוצר).
 *
 * הקובץ מופרד בתו '~' ומקודד בקידוד עברי ישן (IBM/DOS-862). כל שורה מתחילה
 * בקוד רשומה (9015 / 9022 / 9023 / 9024 / 9031 / 9050...), והפענוח נעשה לפי
 * טבלת RECORD_PARSERS: קוד רשומה -> פונקציית פענוח.
 *
 * להוספת סוג רשומה חדש:
 *   1. כותבים פונקציית parseXXXX(fields, lineNo, result) שקוראת את השדות
 *      ומוסיפה את הנתונים ל-result (אפשר להוסיף ל-result אוסף חדש משלה).
 *   2. רושמים אותה ב-RECORD_PARSERS עם מספר השדות המינימלי.
 * שורות עם קוד שאינו רשום מדולגות בשקט (כמו עד היום).
 *
 * מבנה שורת 9050 - תקופות עבודה (לפי סדר השדות בקובץ):
 *     0: קוד רשומה (9050)
 *     1: לא ידוע
 *     2: לא ידוע
 *     3: מספר זהות
 *     4: לא ידוע
 *     5: קוד סוג תקופה
 *     6: תאריך התחלה (DDMMYYYY)
 *     7: תאריך סיום (DDMMYYYY)
 *     8: אורך שירות בחודשים כפול 100 (למשל 01200 = 12 חודשים)
 *     9: קוד סוג זכויות
 *    10: היקף משרה כפול 1000 (למשל 0667 = 0.667)
 */
import fs from "node:fs";
import iconv from "iconv-lite";

/** נרמול מספר זהות: הסרת רווחים ואפסים מובילים (ב-PDF מודפס ללא אפס מוביל). */
export function normalizeId(raw) {
  return raw.trim().replace(/^0+/, "");
}

/** פענוח בייטים של קובץ DAT בקידוד DOS-862 (כל 256 הבייטים מוגדרים בו). */
export function decodeDat(buf) {
  return iconv.decode(buf, "cp862");
}

/** DDMMYYYY -> DD/MM/YYYY לתצוגה. */
export function fmtDate(ddmmyyyy) {
  if (ddmmyyyy && ddmmyyyy.length === 8) {
    return `${ddmmyyyy.slice(0, 2)}/${ddmmyyyy.slice(2, 4)}/${ddmmyyyy.slice(4)}`;
  }
  return ddmmyyyy;
}

/** המרת שדה מספרי; זורק שגיאה על ערך לא מספרי (כמו int() בפייתון). */
function toInt(s) {
  const t = s.trim();
  if (!/^-?\d+$/.test(t)) {
    throw new Error(`ערך לא מספרי: '${s}'`);
  }
  return parseInt(t, 10);
}

/** רשומת 9050: תקופת עבודה אחת - נוספת ל-result.periodsById. */
function parse9050(fields, lineNo, result) {
  const period = {
    idNumber: normalizeId(fields[3]),
    sugTkufa: toInt(fields[5]),
    start: fields[6].trim(),
    end: fields[7].trim(),
    months: toInt(fields[8]) / 100.0,
    sugZchuyot: toInt(fields[9]),
    heikef: toInt(fields[10]) / 1000.0,
    lineNumber: lineNo,
  };
  (result.periodsById[period.idNumber] ??= []).push(period);
}

// קוד רשומה -> { מספר שדות מינימלי, פונקציית פענוח }.
// רשומות נוספות (9015, 9022, 9023...) יתווספו כאן כשיהיה מידע על המבנה שלהן.
const RECORD_PARSERS = {
  "9050": { minFields: 11, parse: parse9050 },
};

export function parseDatBytes(buf) {
  const result = { periodsById: {}, warnings: [], errors: [] };
  const text = decodeDat(buf);
  const lines = text.split(/\r\n|\r|\n/);
  for (let i = 0; i < lines.length; i++) {
    const lineNo = i + 1;
    const line = lines[i];
    const code = line.slice(0, line.indexOf("~"));
    const recordParser = RECORD_PARSERS[code];
    if (!recordParser) continue;
    const fields = line.replace(/\s+$/, "").split("~");
    if (fields.length < recordParser.minFields) {
      result.errors.push(
        `שורה ${lineNo}: שורת ${code} עם ${fields.length} שדות בלבד ` +
        `(נדרשים ${recordParser.minFields}) - השורה דולגה`
      );
      continue;
    }
    try {
      recordParser.parse(fields, lineNo, result);
    } catch (exc) {
      result.errors.push(`שורה ${lineNo}: ערך לא מספרי בשורת ${code} (${exc.message}) - השורה דולגה`);
    }
  }

  checkDuplicatePeriods(result);
  return result;
}

/** בדיקת שפיות: לא אמורות להיות שתי תקופות עם אותם תאריכים לאותה ת"ז. */
function checkDuplicatePeriods(result) {
  for (const [idNumber, periods] of Object.entries(result.periodsById)) {
    const seen = new Map();
    for (const p of periods) {
      const key = `${p.start}|${p.end}`;
      if (seen.has(key)) {
        result.warnings.push(
          `ת"ז ${idNumber}: נמצאו שתי תקופות עם אותם תאריכים ` +
          `(${fmtDate(p.start)} - ${fmtDate(p.end)}, שורות ${seen.get(key)} ו-${p.lineNumber}). ` +
          `ההשוואה עבור התקופה הזו עלולה להיות שגויה.`
        );
      } else {
        seen.set(key, p.lineNumber);
      }
    }
  }
}

export function parseDatFile(path) {
  return parseDatBytes(fs.readFileSync(path));
}
